test/replication-luatest/bootstrap_strategy_test.lua

```diff
@@ -179,3 +181,249 @@ g_auto.test_sync_waits_for_all_connected = function(cg)
+    cg.replica_set = replica_set:new{}
+    cg.server1 = cg.replica_set:build_and_add_server{
+        alias = 'server1',
+        box_cfg = {
+            replication_timeout = 0.1,
+            bootstrap_strategy = 'config',
+            bootstrap_leader = server.build_listen_uri('server1'),
+            replication = {
+                server.build_listen_uri('server2'),
+            },
+        },
+    }
+    cg.server2 = cg.replica_set:build_and_add_server{
+        alias = 'server2',
+        box_cfg = {
+            replication_timeout = 0.1,
+        },
+    }
+end)
+
+g_config.test_replication_without_bootstrap_leader = function(cg)
+    cg.replica_set:start{wait_until_ready = false}
+    local logfile = fio.pathjoin(cg.server1.workdir, 'server1.log')
+    t.helpers.retrying({}, function()
+        t.assert(server:grep_log(no_leader_msg, nil, {filename = logfile}))
+    end)
+end
+
+g_config.after_test('test_replication_without_bootstrap_leader', function(cg)
+    -- Server1 dies and doesn't need to be stopped.
+    cg.server2:stop()
+end)
+
+local set_log_before_cfg = [[
+    local logfile = require('fio').pathjoin(
+        os.getenv('TARANTOOL_WORKDIR'),
+        os.getenv('TARANTOOL_ALIAS') .. '.log'
+    )
+    require('log').cfg{log = logfile}
+]]
+
+g_config.before_test('test_no_leader', function(cg)
+    cg.replica_set = replica_set:new{}
+    cg.server1 = cg.replica_set:build_and_add_server{
+        alias = 'server1',
+        box_cfg = {
+            replication_timeout = 0.1,
+            bootstrap_strategy = 'config',
+            bootstrap_leader = nil,
+            replication = server.build_listen_uri('server1'),
+        },
+        env = {
+            ['TARANTOOL_RUN_BEFORE_BOX_CFG'] = set_log_before_cfg,
+        },
+    }
+end)
+
+g_config.test_no_leader = function(cg)
+    cg.replica_set:start{wait_until_ready = false}
+    local logfile = fio.pathjoin(cg.server1.workdir, 'server1.log')
+    local empty_leader_msg = "the option can't be empty when bootstrap " ..
+                             "strategy is 'config'"
+    t.helpers.retrying({}, function()
+        t.assert(server:grep_log(empty_leader_msg, nil, {filename = logfile}))
+    end)
+end
+
+g_config.before_test('test_single_leader', function(cg)
+    cg.replica_set = replica_set:new{}
+    cg.server1 = cg.replica_set:build_and_add_server{
+        alias = 'server1',
+        box_cfg = {
+            replication_timeout = 0.1,
+            bootstrap_strategy = 'config',
+            bootstrap_leader = server.build_listen_uri('server1'),
+            replication = server.build_listen_uri('server1'),
+        },
+    }
+end)
+
+g_config.test_single_leader = function(cg)
+    cg.replica_set:start()
+    cg.server1:exec(function()
+        local t = require('luatest')
+        t.assert_equals(box.info.status, 'running', 'server is working')
+    end)
+end
+
+g_config.after_test('test_single_leader', function(cg)
+    cg.server1:stop()
+end)
```

Comment:
Up to you. And to others too. I just want to emphasize how extremely complicated it is to read test cases when they are scattered across 3 separate functions. `before_test` and `after_test` are cancer.